src/Shared/RazorShared/MemoryCache.cs

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    // We've created our own MemoryCache here, ideally we would use the one in Microsoft.Extensions.Caching.Memory,
+    // but until we update O# that causes an Assembly load problem.
+    internal class MemoryCache<TKey, TValue>
+    {
+        private static readonly int SizeLimit = 1500;
```

Comment:
Large cache size of 1500 `TagHelperDescriptors` done to ensure it works well for large projects (where these changes are most beneficial).

This should be at most around 3MB:

224 serialized taghelpers are 390KB => 390/224*1500/1024 = 2.55 + CacheEntry overhead ~= 3MB.

Reply:
Given the  reasoning for the size i s based on TagHelpers and you've  made this memory  cache  configurable for  more then just TagHelpers  i'd recommend taking  this  size limit in through the constructor and having  the TagHelperDescriptorCache set it to 1500

Reply:
I was thinking of making this TagHelper specific through the following changes:

- SizeLimit of 1500 to start
- At 1500 perform `Compact` _but_ only compact entries which haven't been used in the past ~15 seconds (we already store the `LastAccess` time. If all 1500 entries have been used recently, then expand the size by 500. 
- The purpose of this is we don't have a significant perf drop off at 1500 entries.

Would you say this additional complexity is worthwhile or too much of an edge case?